### PR TITLE
Various MPI fixes

### PR DESCRIPTION
**Pull request by csegarragonz**
In this PR I fix a couple of issues that I ran into whilst reproducing the MPI results.
* Failing test and workaround for #266. The current solution is only a patch, and we should address it properly in a separate PR.
* Propagate `UNDERFULL` topology hint in MPI to deliberately missplace MPI functions for a migration opportunity to appear.
* Fix race condition in MPI migration.

### CODE REVIEW

File: tests/dist/mpi/test_mpi_functions.cpp [hunk 1]
```diff
@@ -45,6 +45,29 @@ TEST_CASE_METHOD(MpiDistTestsFixture, "Test MPI all to all", "[mpi]")
     checkAllocationAndResult(req);
 }
 
+TEST_CASE_METHOD(MpiDistTestsFixture, "Test MPI all to all many times", "[mpi]")
```
Comment (csegarragonz) [inline tests/dist/mpi/test_mpi_functions.cpp:48]: This distributed test fails in the current `main` branch.

File: src/scheduler/MpiWorld.cpp [hunk 2]
```diff
@@ -1571,7 +1577,31 @@ void MpiWorld::prepareMigration(
...
             assert(hostForRank.at(m.msg().mpirank()) == m.srchost());
+            // Update the host for this rank. We only update the positions of
+            // the to-be migrated ranks, avoiding race conditions with not-
+            // migrated ranks
             hostForRank.at(m.msg().mpirank()) = m.dsthost();
+            // Update the ranks for host. This structure is used when doing
+            // collective communications by all ranks. At this point, all non-
+            // leader ranks will be hitting a barrier, for which they don't
+            // need the ranks for host map, therefore it is safe to modify it
+            if (m.dsthost() == thisHost && m.msg().mpirank() < localLeader) {
```
Comment (csegarragonz) [inline src/scheduler/MpiWorld.cpp:1588]: We used to call `initLocalRemote()` here, unfortunately that method is too coarse and was clearing the `rankForHost` structure before filling it again.

This was causing a race condition between migrated and non-migrated ranks.

File: src/scheduler/ExecGraph.cpp [hunk 3]
```diff
@@ -70,6 +71,45 @@ std::vector<std::string> getMpiRankHostsFromExecGraph(const ExecGraph& graph)
     return getMpiRankHostsFromExecGraphNode(graph.rootNode);
 }
 
+void getMigratedMpiRankHostsFromExecGraph(const ExecGraph& graph,
```
Comment (csegarragonz) [inline src/scheduler/ExecGraph.cpp:74]: Moved this method here so I can use it for Faasm tests as well.

File: src/scheduler/ExecGraph.cpp [hunk 4]
```diff
@@ -70,6 +71,45 @@ std::vector<std::string> getMpiRankHostsFromExecGraph(const ExecGraph& graph)
...
+{
+    std::queue<faabric::scheduler::ExecGraphNode> nodeList;
+    nodeList.push(graph.rootNode);
+    while (!nodeList.empty()) {
+        // Process the node at the front
+        auto node = nodeList.front();
+        int returnValue = node.msg.returnvalue();
+        int rank = node.msg.mpirank();
+        std::string executedHost = node.msg.executedhost();
+        if (returnValue == 0) {
```
Comment (Shillaker) [inline src/scheduler/ExecGraph.cpp:86]: There's 26 lines and up to 3 levels of nesting in this code block, but no linebreaks. A bit of whitespace would make it easier to parse.

File: include/faabric/scheduler/ExecGraph.h [hunk 5]
```diff
@@ -24,6 +24,10 @@ std::set<std::string> getExecGraphHosts(const ExecGraph& graph);
 
 std::vector<std::string> getMpiRankHostsFromExecGraph(const ExecGraph& graph);
 
+void getMigratedMpiRankHostsFromExecGraph(const ExecGraph& graph,
+                                          std::vector<std::string>& hostsBefore,
+                                          std::vector<std::string>& hostsAfter);
```
Comment (Shillaker) [inline include/faabric/scheduler/ExecGraph.h:29]: Passing vectors to be filled is a slightly unintuitive interface for a `getXYZ` method. If you're populating a vector you could call it `populateXYZ` or `appendXYZ`, but instead this should really return the two lists. Returning a `std::pair` of vectors seems more intuitive.

File: src/scheduler/MpiWorld.cpp [hunk 6]
```diff
@@ -123,6 +123,12 @@ void MpiWorld::create(faabric::Message& call, int newId, int newSize)
             msg.set_cmdline(thisRankMsg->cmdline());
             msg.set_inputdata(thisRankMsg->inputdata());
             msg.set_migrationcheckperiod(thisRankMsg->migrationcheckperiod());
+            // To run migration experiments easily, we may want to propagate
+            // the `UNDERFULL` topology hint. In general however, we don't
```
Comment (Shillaker) [inline src/scheduler/MpiWorld.cpp:127]: We don't use markup syntax in inline comments like this as it'll never be rendered as markup.

File: src/scheduler/MpiWorld.cpp [hunk 7]
```diff
@@ -123,6 +123,12 @@ void MpiWorld::create(faabric::Message& call, int newId, int newSize)
...
+            // need to propagate this field
+            if (thisRankMsg->topologyhint() == "UNDERFULL") {
+                msg.set_topologyhint(thisRankMsg->topologyhint());
+            }
```
Comment (Shillaker) [inline src/scheduler/MpiWorld.cpp:131]: It's not enforced in the style guide but after a stand-alone `if` I'd expect a linebreak.

File: tests/dist/mpi/test_mpi_functions.cpp [hunk 8]
```diff
@@ -45,6 +45,29 @@ TEST_CASE_METHOD(MpiDistTestsFixture, "Test MPI all to all", "[mpi]")
...
+        // Set up this host's resources
+        setLocalSlots(nLocalSlots, worldSize);
+        auto req = setRequest("alltoall");
+
+        // Call the functions
+        sch.callFunctions(req);
+
+        checkAllocationAndResult(req);
+
+        SLEEP_MS(500);
```
Comment (Shillaker) [inline tests/dist/mpi/test_mpi_functions.cpp:65]: Why does this need to sleep? As it's running 50 times and sleeping for half a second each time it's adding 25s of sleeping to the test run.

File: tests/dist/mpi/test_mpi_functions.cpp [hunk 9]
```diff
@@ -45,6 +45,29 @@ TEST_CASE_METHOD(MpiDistTestsFixture, "Test MPI all to all", "[mpi]")
...
 }
 
+TEST_CASE_METHOD(MpiDistTestsFixture, "Test MPI all to all many times", "[mpi]")
+{
+    int numRuns = 50;
+    int oldNumLocalSlots = nLocalSlots;
+    nLocalSlots = 4;
+    int worldSize = 8;
+    for (int i = 0; i < numRuns; i++) {
+        SPDLOG_INFO("Starting run {}/{}", i + 1, numRuns);
```
Comment (Shillaker) [inline tests/dist/mpi/test_mpi_functions.cpp:55]: I'm not sure this needs to be `INFO`, probably `DEBUG`.

**Review by Shillaker — CHANGES_REQUESTED** [review]

File: src/scheduler/MpiWorld.cpp [hunk 10]
```diff
@@ -1571,7 +1578,31 @@ void MpiWorld::prepareMigration(
...
+            // leader ranks will be hitting a barrier, for which they don't
+            // need the ranks for host map, therefore it is safe to modify it
+            if (m.dsthost() == thisHost && m.msg().mpirank() < localLeader) {
+                SPDLOG_WARN("Changing local leader {} -> {}",
+                            localLeader,
+                            m.msg().mpirank());
+                localLeader = m.msg().mpirank();
+                ranksForHost[m.dsthost()].insert(
+                  ranksForHost[m.dsthost()].begin(), m.msg().mpirank());
+            }
```
Comment (Shillaker) [inline src/scheduler/MpiWorld.cpp:1596]: Linebreak following an if

File: src/scheduler/MpiWorld.cpp [hunk 11]
```diff
@@ -1571,7 +1578,31 @@ void MpiWorld::prepareMigration(
         for (int i = 0; i < pendingMigrations->migrations_size(); i++) {
             auto m = pendingMigrations->mutable_migrations()->at(i);
             assert(hostForRank.at(m.msg().mpirank()) == m.srchost());
+            // Update the host for this rank. We only update the positions of
+            // the to-be migrated ranks, avoiding race conditions with not-
+            // migrated ranks
             hostForRank.at(m.msg().mpirank()) = m.dsthost();
+            // Update the ranks for host. This structure is used when doing
```
Comment (Shillaker) [inline src/scheduler/MpiWorld.cpp:1585]: Linebreak preceding a comment.

File: src/scheduler/MpiWorld.cpp [hunk 12]
```diff
@@ -1571,7 +1578,31 @@ void MpiWorld::prepareMigration(
         for (int i = 0; i < pendingMigrations->migrations_size(); i++) {
             auto m = pendingMigrations->mutable_migrations()->at(i);
             assert(hostForRank.at(m.msg().mpirank()) == m.srchost());
+            // Update the host for this rank. We only update the positions of
```
Comment (Shillaker) [inline src/scheduler/MpiWorld.cpp:1581]: Linebreak preceding a comment.

File: src/scheduler/MpiWorld.cpp [hunk 13]
```diff
@@ -123,6 +123,13 @@ void MpiWorld::create(faabric::Message& call, int newId, int newSize)
             msg.set_cmdline(thisRankMsg->cmdline());
             msg.set_inputdata(thisRankMsg->inputdata());
             msg.set_migrationcheckperiod(thisRankMsg->migrationcheckperiod());
+            // To run migration experiments easily, we may want to propagate
```
Comment (Shillaker) [inline src/scheduler/MpiWorld.cpp:126]: Linebreak preceding a comment.

File: src/scheduler/ExecGraph.cpp [hunk 14]
```diff
@@ -70,6 +71,56 @@ std::vector<std::string> getMpiRankHostsFromExecGraph(const ExecGraph& graph)
...
+        // Each function in the execution graph (i.e. MPI rank) that has
+        // finished succesfully has either been migrated or not. Each migrated
+        // rank accounts for two functions: one that has been stopped to be
+        // migrated and one that has finished succesfully.
+        if (returnValue == 0) {
+            // If the function has finished succesfully it is either the second
+            // function for the same rank (i.e. it has been migrated) or not
+            if (hostsBefore.at(rank).empty()) {
+                hostsBefore.at(rank) = executedHost;
+            }
```
Comment (Shillaker) [inline src/scheduler/ExecGraph.cpp:101]: Linebreak following an if

File: src/scheduler/MpiWorld.cpp [hunk 15]
```diff
@@ -1571,7 +1578,31 @@ void MpiWorld::prepareMigration(
...
+            }
+            ranksForHost[m.dsthost()].push_back(m.msg().mpirank());
+            ranksForHost[m.srchost()].erase(
+              std::remove(ranksForHost[m.srchost()].begin(),
+                          ranksForHost[m.srchost()].end(),
+                          m.msg().mpirank()),
+              ranksForHost[m.srchost()].end());
+            if (ranksForHost[m.srchost()].empty()) {
+                ranksForHost.erase(m.srchost());
+            }
```
Comment (Shillaker) [inline src/scheduler/MpiWorld.cpp:1605]: Linebreak following an if/ preceding a comment.

File: src/scheduler/MpiWorld.cpp [hunk 16]
```diff
@@ -1571,7 +1578,31 @@ void MpiWorld::prepareMigration(
...
+                ranksForHost[m.dsthost()].insert(
+                  ranksForHost[m.dsthost()].begin(), m.msg().mpirank());
+            }
+            ranksForHost[m.dsthost()].push_back(m.msg().mpirank());
+            ranksForHost[m.srchost()].erase(
+              std::remove(ranksForHost[m.srchost()].begin(),
+                          ranksForHost[m.srchost()].end(),
+                          m.msg().mpirank()),
+              ranksForHost[m.srchost()].end());
+            if (ranksForHost[m.srchost()].empty()) {
```
Comment (Shillaker) [inline src/scheduler/MpiWorld.cpp:1603]: Linebreak preceding this if as it's following a hard-to-read one-liner

**Review by Shillaker — CHANGES_REQUESTED** [review]
More linebreaks.

**Pull request merged**